### PR TITLE
Add helpers to filter wide Pandas matrices by index / columns

### LINKED ISSUE #120: Add method to mask matrices to certain zones
Add a method into the Pandas and Numpy utils that allows the cutting down of a matrix into a subset of it's data.

This is needed in the data requests work here: https://github.com/Transport-for-the-North/data_requests/issues/22

See the following code in NorMITs Demand for examples of how to do this:
- https://github.com/Transport-for-the-North/NorMITs-Demand/blob/12d99c0da03a1fbb9e95011c09a642df36e0f2b6/normits_demand/utils/pandas_utils.py#L679
- https://github.com/Transport-for-the-North/NorMITs-Demand/blob/12d99c0da03a1fbb9e95011c09a642df36e0f2b6/normits_demand/utils/pandas_utils.py#L761
- https://github.com/Transport-for-the-North/NorMITs-Demand/blob/12d99c0da03a1fbb9e95011c09a642df36e0f2b6/normits_demand/utils/pandas_utils.py#L783

### PR DESCRIPTION
**Pull request by BenTaylor-TfN**
Describe Changes
---
Add function(s) to split a wide pandas matrix by specific IDs easily - dealing with edge cases.
Particularly useful for splitting wide zonal matrices down to specific II, IE, EI, and EE areas.

Needed for data requests tool when cutting down matrices of data to a specific area.

fixes #120 

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `pyproject.toml`

### CODE REVIEW

File: tests/pandas_utils/test_wide_df_handling.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+def fixture_non_square_matrix(request) -> WideMatrixData:
+    """Fixture to create random matrix of set dimensions"""
+    return WideMatrixData(matrix=random_matrix(request.param))
+
+
+# # # TESTS # # #
+class TestGetWideMask:
+    """Tests for get_wide_mask()"""
+
+    # TODO(BT): Add tests for datatypes. Different + fuzzy matching. Hoping some
```
Comment (MattBuckley-TfN) [inline tests/pandas_utils/test_wide_df_handling.py:91]: All these tests seem to be calculating the expected matrix, I think it might be worth adding one or two hardcoded matrix results for testing just to make sure there aren't any potential issues in the numpy code which could hide problems in our tests.

**Review by MattBuckley-TfN — COMMENTED** [review]
I think everything looks good and the functions docs are easy to follow. I have a couple of clarification questions and I think an additional hardcoded test would be useful

File: tests/pandas_utils/test_wide_df_handling.py (hunk 1 above)
Comment (BenTaylor-TfN) [reply tests/pandas_utils/test_wide_df_handling.py:91]: Technically this is covered in the doctests. Are you happy accepting that they cover that purpose, or would you rather they were repeated here?

File: src/caf/toolkit/pandas_utils/wide_df_handling.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+    >>> get_wide_mask(df,select=[0, 1],join_fn=operator.or_)
+    array([[ True,  True,  True,  True],
+           [ True,  True,  True,  True],
+           [ True,  True, False, False],
+           [ True,  True, False, False]])
+    """
+    # Validate input args
+    if select is None:
+        if col_select is None or index_select is None:
```
[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply src/caf/toolkit/pandas_utils/wide_df_handling.py:125]: I think this is a good idea. I want to get this PR wrapped up and closed off, so will add an issue for this, and direct this error towards that issue so people are aware they can request this feature if they want it

File: src/caf/toolkit/pandas_utils/wide_df_handling.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+    >>> wide_matrix_internal_external_report(df,[0, 1, 2],[3, 4])
+              internal  external  total
+    internal      54.0      51.0  105.0
+    external     111.0      84.0  195.0
+    total        165.0     135.0  300.0
+    """
+    # Warn if overlap of internal and external selection
+    if len(overlap := set(int_select) & set(ext_select)):
+        warnings.warn(
```
[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply src/caf/toolkit/pandas_utils/wide_df_handling.py:342]: Good catch. Done!

File: tests/pandas_utils/test_wide_df_handling.py (hunk 1 above)
Comment (MattBuckley-TfN) [reply tests/pandas_utils/test_wide_df_handling.py:91]: Completely forgot about those, I don't think there's need to repeat them

**Review by MattBuckley-TfN — APPROVED** [review]
